File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        slow=self.next(n)
        fast=self.next(self.next(n))
        
        while slow!=fast:
            slow=self.next(slow)
            fast=self.next(self.next(fast))
        
        return fast==1
    
    def next(self, n):
        r=0
        while n>0:
            d=n%10
            n//=10
            r+=d*d
        
        return r

File: 0202-happy-number/0202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int slow=digit(n);
        int fast=digit(digit(n));   
        while(slow!=fast){
            slow=digit(slow);
            fast=digit(digit(fast));
        }
        return slow==1;
    }
    int digit(int n){
        int res=0;
        while(n){
            int cur=n%10;
            res+=cur*cur;
            n/=10;
        }
        return res;
    }
};